websockets/client.py

```diff
@@ -272,129 +273,153 @@ def handshake(self, wsuri, origin=None, available_extensions=None,
+        * ``origin`` sets the Origin HTTP header
+        * ``extensions`` is a list of supported extensions in order of
+          decreasing preference
+        * ``subprotocols`` is a list of supported subprotocols in order of
+          decreasing preference
+        * ``extra_headers`` sets additional HTTP request headers – it can be a
+          mapping or an iterable of (name, value) pairs
+        * ``compression`` is a shortcut to configure compression extensions;
+          by default it enables the "permessage-deflate" extension; set it to
+          ``None`` to disable compression
+
+        :func:`connect` raises :exc:`~websockets.uri.InvalidURI` if ``uri``
+        is invalid and :exc:`~websockets.handshake.InvalidHandshake` if the
+        opening handshake fails.
+
+        """
+        if loop is None:
+            loop = asyncio.get_event_loop()
+
+        # Backwards-compatibility: create_protocol used to be called klass.
+        # In the unlikely event that both are specified, klass is ignored.
+        if create_protocol is None:
+            create_protocol = klass
+
+        if create_protocol is None:
+            create_protocol = WebSocketClientProtocol
+
+        wsuri = parse_uri(uri)
+        if wsuri.secure:
+            kwds.setdefault('ssl', True)
+        elif kwds.get('ssl') is not None:
+            raise ValueError("connect() received a SSL context for a ws:// "
+                             "URI, use a wss:// URI to enable TLS")
+
+        if compression == 'deflate':
+            if extensions is None:
+                extensions = []
+            if not any(
+                extension_factory.name == ClientPerMessageDeflateFactory.name
+                for extension_factory in extensions
+            ):
+                extensions.append(ClientPerMessageDeflateFactory(
+                    client_max_window_bits=True,
+                ))
+        elif compression is not None:
+            raise ValueError("Unsupported compression: {}".format(compression))
+
+        factory = lambda: create_protocol(
+            host=wsuri.host, port=wsuri.port, secure=wsuri.secure,
+            timeout=timeout, max_size=max_size, max_queue=max_queue,
+            read_limit=read_limit, write_limit=write_limit,
+            loop=loop, legacy_recv=legacy_recv,
+            origin=origin, extensions=extensions, subprotocols=subprotocols,
+            extra_headers=extra_headers,
         )
-    except Exception:
-        yield from protocol.close_connection(after_handshake=False)
-        raise
 
-    return protocol
+        if kwds.get('sock') is None:
+            host, port = wsuri.host, wsuri.port
+        else:
+            # If sock is given, host and port mustn't be specified.
+            host, port = None, None
+
+        self._wsuri = wsuri
+        self._origin = origin
+
+        # This is a coroutine object.
+        self._creating_connection = loop.create_connection(
+            factory, host, port, **kwds)
```

Comment:
I see that the "split point" between `__init__` and `__await__` occurs just before the first `yield from`.

It seems a bit weird to split the logic that way, but since there's no point in calling `__init__` without `__await__`, I guess that's just pragrmatic.